README


This archive contains the distribution WWW-eNom,
version v1.2.5:

  Interact with eNom, Inc.'s reseller API

This software is Copyright (c) 2013 by Richard Simões.

This is free software, licensed under:

  The MIT (X11) License




Changes
Revision history for WWW-eNom

v1.2.5    2013-11-05 16:11:27 America/Chicago

v1.2.5    2013-11-05 12:00:58 America/Chicago
    - Bumped minimum version of Moo to 1.001000.
    - Fixed test failure from occurring when Moose isn't installed.

v1.2.4    2013-11-04 16:16:02 America/Chicago
    - Required minumum version of Moo and Class::Method::Modifiers.

v1.2.3    2013-11-04 14:08:28 America/Chicago
    - Removed Moose as a testing prerequisite.
    - Pruned down Type::Tiny-related prerequisites.

v1.2.2    2013-11-04 13:53:29 America/Chicago
    - Promoted to full release.

v1.2.1    2013-11-01 21:43:19 America/Chicago
    - Prevented Moo from falling back to Moose by replacing
      __PACKAGE__->meta->add_method with &Class::Method::Modifiers::fresh.

v1.2.0    2013-05-24 15:40:06 America/Chicago
    - Updated command list to reflect eNom's latest version of its reseller API
      version (7.8).
    - Added new, yet-to-be-documented eNom API command (ryeddula).
    - Replace Any::Moose dependency with Moo.

v1.1.3    2013-05-16 10:38:19 America/Chicago
    - Brought tests up to date to match latest eNom response (drzigman).

v1.1.2    2012-06-28 09:20:55 America/Chicago
    - Untabified all code.
    - Replaced Test::Most bundle with only the test modules actually used.
    - Changed license to MIT.

v1.1.1    2012-03-30 20:26:44 America/Chicago
    - Remove parent.pm dependency

v1.1.0    2012-03-06 19:01:26 America/Chicago
    - Removed ParseUtil::Domain dependency in favor of Mozilla::PublicSuffix
    - Fixed POD to show correct license (LGPL v3.0)
    - Other minor POD fixes

v1.0.6    2012-01-16 08:35:24 America/Chicago
    - Increased version requirement of ParseUtil::Domain to 2.10

v1.0.5    2012-01-06 21:21:20 America/Chicago
    - Removed "re" pragma option introduced in Perl v5.14
    - Add deprecation notice to Net::eNom

v1.0.4    2012-01-05 18:12:08 America/Chicago
    - Fixed failing test due to contact email change on cpan.org whois
    - Further reduced depedencies
    - Traded XML::Simple for XML::LibXML::Simple

v1.0.3    2011-12-31 04:29:09 America/Chicago
    - Fixed pod
    - Removed unneeded dependencies

v1.0.2    2011-12-23 21:33:19 America/Chicago
    - Fixed test failures caused by undocument eNom API changes
    - Moved distribution from Net-eNom to WWW-eNom
    - Changed license to LGPL 3.0

v1.0.1    2011-10-23 01:21:53 America/Chicago
    - Added Mouse version requirement to prevent failing compile test in Perl v5.15

v1.0.0    2011-10-03 23:17:37 America/Chicago
    - Renamed main module to WWW::eNom
        -Net::eNom is now a thin wrapper around WWW::eNom
    - Normalized version number
    - Improved test coverage
    - Added version requirement for ParseUtil::Domain to account for new eNom TLD offerings

0.032     2011-09-20
    - Here-and-there whitespace cleanup
    - Fixed version number
    - Taken out of trial mode

v0.3.1    2011-09-19
    - Added direct access to the eNom response type parameter
    - Fixed spelling in comments
    - Removed needless recommended prereq

v0.3.0    2011-09-18
    - Tested and built with with Dist::Zilla
        - POD-related tests are now release tests
    - Added support for eNom API version 6.4
    - Mo(?:o|u)sified

0.02      2009-12-22
    Don't separate "Domain" when the command actually wants "Domain"

0.01      2009-12-22
    First version, released on an unsuspecting world.


lib/WWW/eNom.pm
package WWW::eNom;

use v5.8.9;
use strict;
use warnings;
use utf8;
use Moo 1.001000;
use Type::Tiny 0.032 ();
use Type::Utils qw(class_type subtype as where message);
use Types::Standard qw(Bool Str);
use Carp qw(croak);
use Mozilla::PublicSuffix 0.001016 qw(public_suffix);
use URI 1.60;

our $VERSION = 'v1.2.5'; # VERSION
# ABSTRACT: Interact with eNom, Inc.'s reseller API

with 'WWW::eNom::Role::Commands';

# Supported response types:
my @response_types = qw(xml_simple xml html text);

my $URIObject = class_type({ class => 'URI' })->plus_coercions(
    Str, sub { URI->new($_) }
);

my $eNomResponseType = subtype as Str,
    where {
        my $type = $_;
        { $type eq $_ and return 1 for @response_types; 0 }
    },
    message { 'response_type must be one of: ' . join ', ', @response_types };

has username => (
    isa      => Str,
    is       => 'ro',
    required => 1
);

has password => (
    isa      => Str,
    is       => 'ro',
    required => 1
);

has test => (
    isa     => Bool,
    is      => 'ro',
    default => 0
);

has response_type => (
    isa     => $eNomResponseType,
    is      => 'ro',
    default => 'xml_simple'
);

has _uri => (
    isa     => $URIObject,
    is      => 'ro',
    coerce  => $URIObject->coercion,
    lazy    => 1,
    default => \&_default__uri,
);

sub _make_query_string {
    my ($self, $command, %opts) = @_;
    my $uri = $self->_uri;
    if ( $command ne "CertGetApproverEmail" && exists $opts{Domain} ) {
        my $domain = delete $opts{Domain};
        # Look for an eNom wildcard TLD:
        my $wildcard_tld = qr{\.([*12@]+)$}x;
        my ($subbed_tld) = $domain =~ $wildcard_tld
            and $domain =~ s/$wildcard_tld/.com/x;
        my $suffix = eval { public_suffix($domain) }
            or croak "Domain name, $domain, does not look like a valid domain.";


        # Finally, add in the neccesary API arguments:
        my ($sld) = $domain =~ /^(.+)\.$suffix$/x;
        $suffix = $subbed_tld if $subbed_tld;
        @opts{qw(SLD TLD)} = ($sld, $suffix);
    }

    my $response_type = $self->response_type eq 'xml_simple'
        ? 'xml'
        : $self->response_type;

    $uri->query_form(
        command      => $command,
        uid          => $self->username,
        pw           => $self->password,
        responseType => $response_type,
        %opts
    );

    return $uri;
}

sub _default__uri {
    my $self = shift;
    my $subdomain = $self->test ? 'resellertest' : 'reseller';
    return URI->new("http://$subdomain.enom.com/interface.asp");
}

1;

=encoding utf8

=head1 NAME

WWW::eNom - Interact with eNom, Inc.'s reseller API

=head1 SYNOPSIS

    use WWW::eNom;

    my $enom = WWW::eNom->new(
        username      => "resellid",
        password      => "resellpw",
        response_type => "xml_simple",
        test          => 1 );
    $enom->AddToCart(
        EndUserIP => "1.2.3.4",
        ProductType => "Register",
        SLD => "myspiffynewdomain",
        TLD => "com" );
    ...

=head1 METHODS

=head2 new

Constructs a new object for interacting with the eNom API. If the
"test" parameter is given, then the API calls will be made to the test
server instead of the live one.

As of v0.3.1, an optional "response_type" parameter is supported. For the sake
of backward compatibility, the default is "xml_simple"; see below for an
explanation of this response type. Use of any other valid option will lead to
the return of string responses straight from the eNom API. These options are:

=for Pod::Coverage username password response_type test

=over

=item * xml

=item * html

=item * text

=back

=head2 AddBulkDomains (and many others)

    my $response = $enom->AddBulkDomains(
        ProductType => "register",
        ListCount   => 1,
        SLD1        => "myspiffynewdomain",
        TLD1        => "com",
        UseCart     => 1 );

Performs the specified command - see the
L<eNom API users guide|http://www.enom.com/APICommandCatalog/> for the commands
and their arguments.

For convenience, if you pass the "Domain" argument, it will be split
into "SLD" and "TLD"; that is, you can say

    my $response = $enom->Check( SLD => "myspiffynewdomain", TLD => "com" );

or

    my $response = $enom->Check( Domain => "myspiffynewdomain.com" );

The default return value is a Perl hash (via L<XML::Simple>) representing the
response XML from the eNom API; the only differences are

=over 3

=item *

The "errors" key returns an array instead of a hash

=item *

"responses" returns an array of hashes

=item *

Keys which end with a number are transformed into an array

=back

So for instance, a command C<Check( Domain => "enom.@" )> (the "@" means
"com, net, org, biz, us") might return:

    {
        Domain  => [qw(enom.com enom.net enom.org)],
        Command => "CHECK",
        RRPCode => [qw(211 211 211)],
        RRPText => [
            "Domain not available",
            "Domain not available",
            "Domain not available"
        ]
    }

You will need to read the API guide to check whether to expect responses
in "RRPText" or "responses"; it's not exactly consistent.

=head1 RELEASE NOTE

As of v1.0.0, this module has been renamed to WWW::eNom. Net::eNom is now a thin
wrapper to preserve backward compatibility.

=head1 AUTHOR

Richard Simões, C<< <rsimoes AT cpan DOT org> >>

Original version by Simon Cozens.

=head1 COPYRIGHT & LICENSE

Copyright © 2013 Richard Simões. This module is released under the terms of the
B<MIT License> and may be modified and/or redistributed under the same or any
compatible license.


lib/WWW/eNom/Role/Commands.pm
package WWW::eNom::Role::Commands;

use Moo::Role;
use strict;
use warnings;
use utf8;

use Class::Method::Modifiers 2.04 qw(fresh);
use HTTP::Tiny 0.031;
use XML::LibXML::Simple 0.91 qw(XMLin);

our $VERSION = 'v1.2.5'; # VERSION

requires "_make_query_string";

=begin Pod::Coverage

 \w+

=end Pod::Coverage

=cut

# Create methods to support eNom API version 7.8:
my @commands = qw(
    AddBulkDomains AddContact AddDomainFolder AddDomainHeader AddHostHeader
    AddToCart AdvancedDomainSearch AM_AutoRenew AM_Configure AM_GetAccountDetail
    AM_GetAccounts AM_Renew AssignToDomainFolder AuthorizeTLD
    CalculateAllHostPackagePricing CalculateHostPackagePricing CancelHostAccount
    CertChangeApproverEmail CertConfigureCert CertGetApproverEmail
    CertGetCertDetail CertGetCerts CertModifyOrder CertParseCSR CertReissueCert
    CertPurchaseCert CertResendApproverEmail CertResendFulfillmentEmail Check
    CheckLogin CheckNSStatus CommissionAccount Contacts CreateAccount
    CreateHostAccount CreateSubAccount DeleteAllPOPPaks DeleteContact
    DeleteCustomerDefinedData DeleteDomainFolder DeleteDomainHeader
    DeleteFromCart DeleteHostedDomain DeleteHostHeader DeleteNameServer
    DeletePOP3 DeletePOPPak DeleteRegistration DeleteSubaccount DisableFolderApp
    DisableServices EnableFolderApp EnableServices Extend Extend_RGP
    ExtendDomainDNS Forwarding GetAccountInfo GetAccountPassword
    GetAccountValidation GetAddressBook GetAgreementPage GetAllAccountInfo
    GetAllDomains GetAllHostAccounts GetAllResellerHostPricing GetBalance
    GetCartContent GetCatchAll GetCerts GetConfirmationSettings GetContacts
    GetCusPreferences GetCustomerDefinedData GetCustomerPaymentInfo GetDNS
    GetDNSStatus GetDomainCount GetDomainExp GetDomainFolderDetail
    GetDomainFolderList GetDomainHeader GetDomainInfo GetDomainNameID GetDomains
    GetDomainServices GetDomainSLDTLD GetDomainSRVHosts GetDomainStatus
    GetDomainSubServices GetDotNameForwarding GetExpiredDomains GetExtAttributes
    GetExtendInfo GetFilePermissions GetForwarding GetGlobalChangeStatus
    GetGlobalChangeStatusDetail GetHomeDomainList GetHostAccount GetHostAccounts
    GetHostHeader GetHosts GetIDNCodes GetIPResolver GetMailHosts GetMetaTag
    GetNews GetOrderDetail GetOrderList GetPasswordBit GetPOP3 GetPOPExpirations
    GetPOPForwarding GetProductNews GetProductSelectionList GetRegHosts
    GetRegistrationStatus GetRegLock GetRenew GetReport GetResellerHostPricing
    GetResellerInfo GetServiceContact GetSPFHosts GetStorageUsage
    GetSubAccountDetails GetSubAccountPassword GetSubAccounts
    GetSubaccountsDetailList GetTLDList GetTransHistory GetWebHostingAll
    GetWhoisContact GetWPPSInfo GM_CancelSubscription GM_CheckDomain
    GM_GetCancelReasons GM_GetControlPanelLoginURL GM_GetRedirectScript
    GM_GetStatuses GM_GetSubscriptionDetails GM_GetSubscriptions
    GM_ReactivateSubscription GM_RenewSubscription GM_UpdateBillingCycle
    GM_UpdateSubscriptionDetails HostPackageDefine HostPackageDelete
    HostPackageModify HostPackageView HostParkingPage InsertNewOrder
    IsFolderEnabled ListDomainHeaders ListHostHeaders ListWebFiles
    MetaBaseGetValue MetaBaseSetValue ModifyDomainHeader ModifyHostHeader
    ModifyNS ModifyNSHosting ModifyPOP3 MySQL_GetDBInfo NameSpinner
    NM_CancelOrder NM_ExtendOrder NM_GetPremiumDomainSettings
    NM_GetSearchCategories NM_ProcessOrder NM_Search
    NM_SetPremiumDomainSettings ParseDomain PE_GetCustomerPricing
    PE_GetDomainPricing PE_GetPOPPrice PE_GetProductPrice PE_GetResellerPrice
    PE_GetRetailPrice PE_GetRetailPricing PE_GetRocketPrice PE_GetTLDID
    PE_SetPricing PP_CancelSubscription PP_CheckUpgrade PP_GetCancelReasons
    PP_GetControlPanelLoginURL PP_GetStatuses PP_GetSubscriptionDetails
    PP_GetSubscriptions PP_ReactivateSubscription PP_UpdateSubscriptionDetails
    PP_ValidatePassword Portal_GetToken PreConfigure Purchase PurchaseHosting
    PurchasePOPBundle PurchasePreview PurchaseServices PushDomain
    RC_CancelSubscription RC_FreeTrialCheck RC_GetLoginToken
    RC_GetSubscriptionDetails RC_GetSubscriptions RC_RebillSubscription
    RC_ResetPassword RC_SetBillingCycle RC_SetPassword RC_SetSubscriptionDomain
    RC_SetSubscriptionName RefillAccount RegisterNameServer RemoveTLD
    RemoveUnsyncedDomains RenewPOPBundle RenewServices RP_CancelAccount
    RP_GetAccountDetail RP_GetAccounts RP_GetUpgradeOptions RP_GetUpgradePrice
    RP_RebillAccount RP_UpdateAccountName RP_UpgradeAccount RP_ValidateEmail
    RPT_GetReport SendAccountEmail ServiceSelect SetCatchAll
    SetCustomerDefinedData SetDNSHost SetDomainSRVHosts SetDomainSubServices
    SetDotNameForwarding SetFilePermissions SetHosts SetIPResolver SetPakRenew
    SetPassword SetPOPForwarding SetRegLock SetRenew SetResellerServicesPricing
    SetResellerTLDPricing SetSPFHosts SetUpPOP3User SL_AutoRenew SL_Configure
    SL_GetAccountDetail SL_GetAccounts StatusDomain SubAccountDomains
    SynchAuthInfo TEL_AddCTHUser TEL_GetCTHUserInfo TEL_GetCTHUserList
    TEL_GetPrivacy TEL_IsCTHUser TEL_UpdateCTHUser TEL_UpdatePrivacy
    TLD_AddWatchlist TLD_DeleteWatchlist TLD_GetTLD TLD_GetWatchlist
    TLD_GetWatchlistTlds TP_CancelOrder TP_CreateOrder TP_GetDetailsByDomain
    TP_GetOrder TP_GetOrderDetail TP_GetOrderReview TP_GetOrdersByDomain
    TP_GetOrderStatuses TP_GetTLDInfo TP_ResendEmail TP_ResubmitLocked
    TP_SubmitOrder TP_UpdateOrderDetail TS_AutoRenew TS_Configure
    TS_GetAccountDetail TS_GetAccounts UpdateAccountInfo UpdateAccountPricing
    UpdateCart UpdateCusPreferences UpdateDomainFolder UpdateExpiredDomains
    UpdateHostPackagePricing UpdateMetaTag UpdateNameServer
    UpdateNotificationAmount UpdatePushList UpdateRenewalSettings
    ValidatePassword WBLConfigure WBLGetCategories WBLGetFields WBLGetStatus
    WebHostCreateDirectory WebHostCreatePOPBox WebHostDeletePOPBox
    WebHostGetCartItem WebHostGetOverageOptions WebHostGetOverages
    WebHostGetPackageComponentList WebHostGetPackageMinimums WebHostGetPackages
    WebHostGetPOPBoxes WebHostGetResellerPackages WebHostGetStats
    WebHostHelpInfo WebHostSetCustomPackage WebHostSetOverageOptions
    WebHostUpdatePassword WebHostUpdatePOPPassword WSC_GetAccountInfo
    WSC_GetAllPackages WSC_GetPricing WSC_Update_Ops XXX_GetMemberId
    XXX_RemoveMemberId XXX_SetMemberId
    GetDomainAuthInfo
);

my $ua = HTTP::Tiny->new;
for my $command (@commands) {
    fresh $command => sub {
        my ($self, @opts) = @_;
        my $uri = $self->_make_query_string($command, @opts);
        my $response = $ua->get($uri)->{content};
        my $response_type = $self->response_type;
        if ( $response_type eq "xml_simple" ) {
            $response = XMLin($response);
            $response->{errors} &&= [ values %{ $response->{errors} } ];
            $response->{responses} &&= $response->{responses}{response};
            $response->{responses} = [ $response->{responses} ]
                if $response->{ResponseCount} == 1;
            foreach my $key ( keys %{$response} ) {
                next unless $key =~ /(.*?)(\d+)$/;
                $response->{$1} = undef if ref $response->{$key};
                $response->{$1}[ $2 - 1 ] = delete $response->{$key};
            }
        }
        return $response;
    };
}

1;


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Interact with eNom, Inc.'s reseller API",
  "dist_author" => [
    "Richard Sim\x{f5}es <rsimoes AT cpan DOT org>"
  ],
  "dist_name" => "WWW-eNom",
  "dist_version" => "v1.2.5",
  "license" => "mit",
  "module_name" => "WWW::eNom",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Class::Method::Modifiers" => "2.04",
    "HTTP::Tiny" => "0.031",
    "Moo" => "1.001000",
    "Mozilla::PublicSuffix" => "0.001016",
    "Type::Tiny" => "0.032",
    "URI" => "1.60",
    "XML::LibXML::Simple" => "0.91",
    "perl" => "v5.8.9"
  },
  "script_files" => [],
  "test_requires" => {
    "File::Spec" => 0,
    "HTML::Parser" => "v3.67.0",
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Deep" => 0,
    "Test::Exception" => 0,
    "Test::More" => 0,
    "Test::Warn" => 0
  }
);


my %fallback_build_requires = (
  "File::Spec" => 0,
  "HTML::Parser" => "v3.67.0",
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.3601",
  "Test::Deep" => 0,
  "Test::Exception" => 0,
  "Test::More" => 0,
  "Test::Warn" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
